### ISSUE TITLE
Add the ability to take snapshots of the physical world.

### ISSUE THREAD

**Issue by sebcrozet** [sebcrozet/nphysics#83]
See the related thread on the [forum](http://users.nphysics.org/t/is-it-possible-to-create-world-snapshots).

**Comment by SirRade** [comment]
Has there been any progress on this?  
It would be quintessential for multiplayer online games.

**Comment by Ralith** [comment]
For most game purposes, exact reproducibility isn't important (especially if you're receiving authoritative updates from a server at high frequency), so you can just synchronize position and velocity by hand.

**Comment by SirRade** [comment]
I agree for the client, but [server reconciliation](http://www.gabrielgambetta.com/client-side-prediction-server-reconciliation.html) requires you to "scroll backwards" in time.

**Comment by Ralith** [comment]
Efficient lag compensation implementations usually don't involve snapshotting the entire world state at every frame, but just reconstructing sufficiently important parts of the past (e.g. player positions). They certainly don't need the ability to restart the simulation from a point in the past, let alone perfectly reproducibly.

**Comment by seivan** [comment]
Depends on what kind of roll back you're looking for. For Counter Strike, they just keep track of transforms and stuff like ammo/weapon/aim and use those for hit-scan rollbacks and they tend to favour the shooter. 

More complex games with different rules regarding who to "favour" will need to rollback much more. Take Overwatch that has projectiles will need to tackle a bit more than just hitscan + transforms. 

Although,  there is nothing preventing you to store the values you deem important for a rollback, and re-simulate with said values seeded on demand it would be nice if this was builtin.

**Comment by Ralith** [comment]
I don't think restarting the entire simulation from the past is generally required or appropriate for hit detection, no matter how you're simulating your projectiles.

**Comment by seivan** [comment]
So what approach do you recommend? Also hit detection was just an example there’s much more to roll back for.
